bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/MarketplaceBundleInstaller.java

```diff
@@ -142,26 +146,25 @@ protected void uninstallBundle(BundleContext bundleContext, String addonId) thro
      * @param bundleContext the {@link BundleContext} to use to look up the bundles
      */
     protected void ensureCachedBundlesAreInstalled(BundleContext bundleContext) {
-        File addonPath = new File(BUNDLE_CACHE_PATH);
-        if (addonPath.exists() && addonPath.isDirectory()) {
-            for (File bundleFile : addonPath.listFiles()) {
-                if (bundleFile.isDirectory()) {
-                    String addonId = "marketplace:" + bundleFile.getName();
-                    if (!isBundleInstalled(bundleContext, addonId)) {
-                        logger.info("Reinstalling missing marketplace bundle: {}", addonId);
-                        try {
-                            installFromCache(bundleContext, addonId);
-                        } catch (MarketplaceHandlerException e) {
-                            logger.warn("Failed reinstalling add-on from cache", e);
-                        }
-                    }
-                }
-                bundleFile.delete();
+        try {
+            if (Files.isDirectory(BUNDLE_CACHE_PATH)) {
+                Files.list(BUNDLE_CACHE_PATH).filter(Files::isDirectory).map(p -> "marketplace:" + p.getFileName())
+                        .filter(addonId -> !isBundleInstalled(bundleContext, addonId)).forEach(addonId -> {
+                            logger.info("Reinstalling missing marketplace bundle: {}", addonId);
+                            try {
+                                installFromCache(bundleContext, addonId);
+                            } catch (MarketplaceHandlerException e) {
+                                logger.warn("Failed reinstalling add-on from cache", e);
+                            }
+                        });
+                // TODO: why is that: bundleFile.delete();
```

Comment:
@ghys Can you comment here? Why is the file deleted from the cache after it has been installed?

Reply:
It's the directory with the topic ID, and since it's not empty, nothing would happen. I think it's an unfortunate leftover, so the best is to remove it.